csrc/scheduler/matmul_utils.cpp

```diff
@@ -416,10 +416,42 @@ std::shared_ptr<MatmulParams> getMatmulHeuristics(
   const auto& roles_map_opt = mma_utils::getTensorsRoles(fusion);
   TORCH_INTERNAL_ASSERT(
       roles_map_opt.isValid(), "Tensor roles map in mma is not valid.");
+
+  // smem_a and smem_b are guaranteed to be re-used for smem_c as long as:
+  //   - they are marked for re-use using promoteReuse
+  //   - they are not aliased by another tensor whose lifetime extends past the
+  //   start of smem_epilogue's.
+  //   - their lifetimes do not overlap smem_epilogue
+  //
+  // We guarantee the first condition by calling tv->promoteReuse() in
+  // scheduleProlog.
+  //
+  // The second condition would only be the case if another smem tensor had the
+  // same indexing and its lifetime did not overlap. This scheduler only uses
+  // smem for these three arrays, so the only candidate for aliasing is C. If C
+  // aliases either A or B, the following expression is still valid.
+  //
+  // The third condition is satisfied in the simple cases where the inputs to
+  // the matmul have only this use. However, it could be violated if a or b has
+  // other uses that get ordered after the matmul; for example when computing
+  // matmul(A, B) + A for square matrices A and B. In that case, the smem tensor
+  // resulting from A->cacheAfter() will be used in both the matmul as well as
+  // the addition that occurs in the epilogue, extending the lifetime such that
+  // it violates the third condition above. In order to avoid errors in these
+  // cases, we check that there is no re-use when there is more than one use of
+  // either a or b. If there are multiple uses we might wind up re-using memory,
+  // but in that case the calculation below will be overly conservative.
+  const auto roles_map = roles_map_opt.getData();
+  TensorView* a = roles_map.at(MatmulRole::INPUT_A).front();
+  TensorView* b = roles_map.at(MatmulRole::INPUT_B).front();
+  bool smem_a_reuse_guaranteed = a->uses().size() == 1;
+  bool smem_b_reuse_guaranteed = b->uses().size() == 1;
   params->use_smem_epilogue = mma_utils::generateSharedMemoryEpilogueHeuristics(
```

Comment:
A minor suggestion:

How about passing `const mma_utils::RolesMap* roles_map` (not the optional, but the real object)?
This will allow us dropping last three parameters as they only depend on roles map. We can access required data from within `mma_utils::generateSharedMemoryEpilogueHeuristics(...)` so all smem related activities are done within this function.

Reply:
Good suggestion. I did that, but found that the tests call `generateSharedMemoryEpilogueHeuristics` directly in a way that was a bit ugly to work around since in some cases like in `test_matmul_sass.cpp` the params are used independently of a `Fusion`; in that case I'd need to either create a `Fusion` just to create some params, or move the call to `generateSharedMemoryEpilogueHeuristics` into `getSASS`. The former is a bit distracting and the latter complicates skipping the epilogue test when smem is not used. So I figured the simplest approach is to just overload the function so you can just pass a `RolesMap`, which is done in `getMatmulHeuristics`, but keep the original also which is used in the tests.